Optimize some Matrix4x4 operations with SSE

EgorBo opened the pull request:
This PR optimizes some `Matrix4x4` operations with SSE (see https://github.com/dotnet/corefx/issues/31425). Some of the operations could also be optimized with AVX but for some reason on my PC it performs worse than SSE ([VZEROUPPER](https://github.com/dotnet/coreclr/issues/18575)? or maybe CPU decreases AVX frequency due to rapid benchmarking?).

Environment:
```ini
.NET Core: .NET Core SDK=3.0.100-alpha1-20180720-2

Windows 10:
   Intel Core i7-8700K CPU 3.70GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores

macOS 10.13:
   Intel Core i7-4980HQ CPU 2.80GHz (Haswell), 1 CPU, 8 logical and 4 physical cores
```

## Matrix4x4.Add (Matrix4x4, Matrix4x4) and Subtract
`Matrix4x4 result = matrix1 + matrix2;`

Windows (Coffee Lake):

|  Method |      Mean | Scaled |
|-------- |----------:|-------:|
| Add_old | 13.353 ns |   1.00 |
| Add_new |  4.486 ns |   0.34 |

macOS (Haswell):

|  Method |      Mean | Scaled |
|-------- |----------:|-------:|
| Add_old | 15.347 ns |   1.00 |
| Add_new |  7.473 ns |   0.49 |


## Matrix4x4.Lerp (Matrix4x4, Matrix4x4, float)
`Matrix4x4 result = Matrix4x4.Lerp(matrix1, matrix2, amount);`

Windows (Coffee Lake):

|   Method |      Mean | Scaled |
|--------- |----------:|-------:|
| Lerp_old | 15.286 ns |   1.00 |
| Lerp_new |  5.365 ns |   0.35 |

macOS (Haswell):

|   Method |      Mean | Scaled |
|--------- |----------:|-------:|
| Lerp_old | 17.047 ns |   1.00 |
| Lerp_new |  7.657 ns |   0.45 |

## Matrix4x4.Multiply (Matrix4x4, Matrix4x4)
`Matrix4x4 result = matrix1 * matrix2;`

Windows (Coffee Lake):

|       Method |      Mean | Scaled |
|------------- |----------:|-------:|
| Multiply_old | 27.146 ns |   1.00 |
| Multiply_new |  7.461 ns |   0.27 |

macOS (Haswell):

|       Method |     Mean | Scaled |
|------------- |---------:|-------:|
| Multiply_old | 32.05 ns |   1.00 |
| Multiply_new | 11.24 ns |   0.35 |

## Matrix4x4.Multiply (Matrix4x4, float)
`Matrix4x4 result = matrix1 * scalar;`

Windows (Coffee Lake):

|               Method |      Mean | Scaled |
|--------------------- |----------:|-------:|
| MultiplyByScalar_old | 12.927 ns |   1.00 |
| MultiplyByScalar_new |  3.284 ns |   0.25 |

macOS (Haswell):

|               Method |      Mean | Scaled |
|--------------------- |----------:|-------:|
| MultiplyByScalar_old | 14.334 ns |   1.00 |
| MultiplyByScalar_new |  5.086 ns |   0.35 |

## Matrix4x4.Negate (Matrix4x4)
`Matrix4x4 result = -matrix1;`

Windows (Coffee Lake):

|     Method |      Mean | Scaled |
|----------- |----------:|-------:|
| Negate_old | 12.932 ns |   1.00 |
| Negate_new |  3.187 ns |   0.25 |

macOS (Haswell):

|     Method |      Mean | Scaled |
|----------- |----------:|-------:|
| Negate_old | 14.877 ns |   1.00 |
| Negate_new |  5.201 ns |   0.35 |

## Matrix4x4.Equals (Matrix4x4)
`bool result = matrix1 == matrix2;`

Windows (Coffee Lake):

|              Method |     Mean |
|-------------------- |---------:|
| Equals_NotEqual_old | 1.742 ns |
| Equals_NotEqual_new | 1.581 ns |
|    Equals_Equal_old | 7.081 ns |
|    Equals_Equal_new | 2.960 ns |

macOS (Haswell):

|              Method |     Mean |
|-------------------- |---------:|
| Equals_NotEqual_old | 3.172 ns |
| Equals_NotEqual_new | 3.022 ns |
|    Equals_Equal_old | 8.180 ns |
|    Equals_Equal_new | 4.618 ns |

## Matrix4x4.Transpose (Matrix4x4)
`Matrix4x4 result = Matrix4x4.Transpose(matrix1);`

Windows (Coffee Lake):

|        Method |      Mean | Scaled |
|-------------- |----------:|-------:|
| Transpose_old | 12.720 ns |   1.00 |
| Transpose_new |  3.156 ns |   0.25 |

macOS (Haswell):

|        Method |      Mean | Scaled |
|-------------- |----------:|-------:|
| Transpose_old | 14.297 ns |   1.00 |
| Transpose_new |  5.548 ns |   0.39 |

Benchmarks: https://gist.github.com/EgorBo/c80a25517245374c8dcdca2af4536ffe

tannergooding commented:
Possibly outside the scope of this PR, but it would be nice if we had some perf tests for `Matrix4x4` (and the other types) in: https://github.com/dotnet/corefx/tree/master/src/System.Numerics.Vectors/tests/Performance

tannergooding commented:
CC. @eerhardt, @ViktorHofer 

Also CC. @benaadams, who has looked at some of this before.

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1786,64 +1787,102 @@ public static Matrix4x4 Transpose(Matrix4x4 matrix)
-            result.M44 = -value.M44;
+            if (Sse.IsSupported)
+            {
+                var zero = Sse.SetAllVector128(0f);
```
Please use `SetZeroVector128` instead

eerhardt commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1786,64 +1787,102 @@ public static Matrix4x4 Transpose(Matrix4x4 matrix)
-            result.M12 = matrix1.M12 + (matrix2.M12 - matrix1.M12) * amount;
-            result.M13 = matrix1.M13 + (matrix2.M13 - matrix1.M13) * amount;
-            result.M14 = matrix1.M14 + (matrix2.M14 - matrix1.M14) * amount;
+                var m1Row = Sse.LoadVector128(&matrix1.M11);
```
Do we need to have `[StructLayout(LayoutKind.Sequential)]` on the `Matrix4x4` type? That way we are assured the members will be laid out in the assumed order?

EgorBo replied:
is not the default layout sequential? 
Also there is an opinion that a better layout for matrices is column-major order (this one is row-major) for popular matrix operations in game development (but of course it's a breaking change and doesn't match DirectX)

tannergooding replied:
The Roslyn Compiler emits `StructLayout=Sequential` by default for struct types. However, last I checked there wasn't anything enforcing this behavior in the language spec.

It would probably be a good idea to explicitly mark the numeric types as Sequential (as we do with types like `Single` and `Int32`).

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1912,35 +1973,68 @@ public static Matrix4x4 Subtract(Matrix4x4 value1, Matrix4x4 value2)
+                Matrix4x4 m = default;
+
+                Sse.Store(&m.M11,
+                    Sse.Add(Sse.Add(Sse.Multiply(Sse.SetAllVector128(value1.M11), Sse.LoadVector128(&value2.M11)),
```
Rather than calling `SetAllVector128` multiple times, it is probably better to do a load row and permute (broadcast in AVX, shuffle in SSE)

EgorBo replied:
btw, it'd be nice to expose `_MM_SHUFFLE` macros to C# for Shuffle.
e.g. `static byte _MM_SHUFFLE(byte fp3, byte fp2, byte fp1, byte fp0) => (byte)(((fp3) << 6) | ((fp2) << 4) | ((fp1) << 2) | ((fp0)));`

tannergooding replied:
I've raised it before. It should go to API review soon

eerhardt replied:
Agreed. @tannergooding and I had the same conclusion during https://github.com/dotnet/machinelearning/pull/562.  However, I didn't see an issue open for this in either corefx or coreclr.  @tannergooding did you ever open one?

Here is what we proposed at the time:
```
 `Sse.Shuffle(op1, op2, Sse.GetShuffleControl(0, 3, 1, 2))` 
Something like that. `GetShuffleControl`, `BuildShuffleControl`, or something
```

EgorBo replied:
So I've tried to use shuffle here but it became slower, maybe I did something wrong? 
![image](https://user-images.githubusercontent.com/523221/44172305-707d1280-a0e5-11e8-915c-acb2f8498bc7.png)

tannergooding replied:
I'll take a look at the assembly and see if I can spot what is going wrong....

tannergooding replied:
Logged a bug to track this here: https://github.com/dotnet/corefx/issues/31825

EgorBo replied:
@tannergooding any update on `_MM_SHUFFLE ` API? Could not find an issue for it

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1949,57 +2043,82 @@ public static Matrix4x4 Multiply(Matrix4x4 value1, Matrix4x4 value2)
-            m.M44 = -value.M44;
-
-            return m;
+            if (Sse.IsSupported)
```
Why is this not just calling the `Negate` method (or vice-versa)?

tannergooding replied:
Same for the operators below

EgorBo replied:
I was afraid Negate call won't be inlined, let me check.

tannergooding replied:
I believe the JIT will inline it so that the inner call is called directly, rather than trying to inline the inner call into the outer call.
If it doesn't, then this sounds like a good JIT bug to log.

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1786,64 +1787,102 @@ public static Matrix4x4 Transpose(Matrix4x4 matrix)
-            result.M14 = matrix1.M14 + (matrix2.M14 - matrix1.M14) * amount;
+                var m1Row = Sse.LoadVector128(&matrix1.M11);
+                var m2Row = Sse.LoadVector128(&matrix2.M11);
+                Sse.Store(&matrix1.M11, Sse.Add(m1Row, Sse.Multiply(Sse.Subtract(m2Row, m1Row), amountVec)));
```
Same for all the operations below.

tannergooding commented:
Also CC. @CarolEidt

EgorBo commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1949,219 +2050,104 @@ public static Matrix4x4 Multiply(Matrix4x4 value1, Matrix4x4 value2)
-                    value1.M31 != value2.M31 || value1.M32 != value2.M32 || value1.M33 != value2.M33 || value1.M34 != value2.M34 ||
-                    value1.M41 != value2.M41 || value1.M42 != value2.M42 || value1.M43 != value2.M43 || value1.M44 != value2.M44);
-        }
+        public static bool operator !=(Matrix4x4 value1, Matrix4x4 value2) => !value1.Equals(value2);
```
I am not sure about this one, will check the jit output

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1786,64 +1790,103 @@ public static Matrix4x4 Transpose(Matrix4x4 matrix)
         /// <param name="value">The source matrix.</param>
         /// <returns>The negated matrix.</returns>
-        public static Matrix4x4 Negate(Matrix4x4 value)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
I'm not sure about this. The code is probably small enough to be inlined in the hardware accelerated case, but not necessarily in the software case.

tannergooding replied:
Same goes for other new `AggressiveInlining` attributes

tannergooding replied:
What is the codegen you are seeing here?

EgorBo replied:
@tannergooding Negate is not inlined 🙁 https://gist.github.com/EgorBo/a6cc52d5523d7a5fcf6fb1adfa5dfbc0 
`Negate_new` is a benchmark method that does `return Matrix4x4X.Negate(m1x);`

EgorBo replied:
thus current numbers are:

|        Method |      Mean |     Error |    StdDev | Scaled |
|-------------- |----------:|----------:|----------:|-------:|
|    Negate_old | 13.385 ns | 0.0781 ns | 0.0693 ns |   1.00 |
|    Negate_new |  3.364 ns | 0.0230 ns | 0.0192 ns |   0.25 |
| Negate_new_op |  3.023 ns | 0.0257 ns | 0.0240 ns |   0.23 |

tannergooding replied:
I'm seeing that for `var negated = Matrix4x4.Negate(Matrix4x4.Identity);` where `public static Matrix4x4 Negate(Matrix4x4 value) => -value`, there is only a single call and it contains the relevant SSE code (there isn't a call, with an immediate nested call).

I'm building CoreCLR locally to get the jit-utils diff and will post sometime after lunch.

EgorBo replied:
um.. I expected `Negate_new()` to just call op_UnaryNegation directly  
![image](https://user-images.githubusercontent.com/523221/44167843-f2663f00-a0d7-11e8-9165-8f2c439dcaa1.png)

tannergooding replied:
Ok, using `jitutils` to get PMI diffs, and for `public static unsafe Matrix4x4 Negate(Matrix4x4 value) => -value;` and `public static unsafe Matrix4x4 operator -(Matrix4x4 value)` containing the actual implementation....

The following two tests (aside from not being "correct"):
```C#
[Fact]
public void Matrix4x4NegateTest()
{
    Matrix4x4 m = GenerateMatrixNumberFrom1To16();
    var actual = Matrix4x4.Negate(m);
    Assert.Equal(m, actual);
}

[Fact]
public void Matrix4x4op_NegateTest()
{
    Matrix4x4 m = GenerateMatrixNumberFrom1To16();
    var actual = -m;
    Assert.Equal(m, actual);
}
```

Both generate a call directly to `call     Matrix4x4:op_UnaryNegation(struct):struct`:
```asm
; Assembly listing for method Matrix4x4Tests:Matrix4x4NegateTest():this
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;* V00 this         [V00    ] (  0,  0   )     ref  ->  zero-ref    this class-hnd
;  V01 loc0         [V01    ] (  3,  3   )  struct (64) [rsp+0x1A8]   do-not-enreg[XSB] must-init addr-exposed
;  V02 loc1         [V02    ] (  2,  2   )  struct (64) [rsp+0x168]   do-not-enreg[XSB] must-init addr-exposed
;  V03 OutArgs      [V03    ] (  1,  1   )  lclBlk (32) [rsp+0x00]  
;  V04 tmp1         [V04,T01] (  2,  4   )  struct (64) [rsp+0x128]   do-not-enreg[SB]
;  V05 tmp2         [V05,T02] (  2,  4   )  struct (64) [rsp+0xE8]   do-not-enreg[SB]
;  V06 tmp3         [V06,T03] (  2,  4   )  struct (64) [rsp+0xA8]   do-not-enreg[SB]
;  V07 tmp4         [V07,T00] (  3,  6   )     ref  ->  rsi         class-hnd exact
;  V08 tmp5         [V08    ] (  4,  8   )  struct (64) [rsp+0x68]   do-not-enreg[XSB] addr-exposed
;  V09 tmp6         [V09,T04] (  2,  4   )    long  ->  rcx        
;  V10 tmp7         [V10    ] (  2,  4   )  struct (64) [rsp+0x28]   do-not-enreg[XSB] addr-exposed
;
; Lcl frame size = 488
G_M53037_IG01:
       push     rdi
       push     rsi
       sub      rsp, 488
       vzeroupper 
       mov      rsi, rcx
       lea      rdi, [rsp+168H]
       mov      ecx, 32
       xor      rax, rax
       rep stosd 
       mov      rcx, rsi
G_M53037_IG02:
       lea      rcx, bword ptr [rsp+1A8H]
       call     Matrix4x4Tests:GenerateMatrixNumberFrom1To16():struct
       vmovdqu  xmm0, qword ptr [rsp+1A8H]
       vmovdqu  qword ptr [rsp+128H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1B8H]
       vmovdqu  qword ptr [rsp+138H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1C8H]
       vmovdqu  qword ptr [rsp+148H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1D8H]
       vmovdqu  qword ptr [rsp+158H], xmm0
       lea      rcx, bword ptr [rsp+168H]
       vmovdqu  xmm0, qword ptr [rsp+128H]
       vmovdqu  qword ptr [rsp+68H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+138H]
       vmovdqu  qword ptr [rsp+78H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+148H]
       vmovdqu  qword ptr [rsp+88H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+158H]
       vmovdqu  qword ptr [rsp+98H], xmm0
       lea      rdx, bword ptr [rsp+68H]
       call     Matrix4x4:op_UnaryNegation(struct):struct
       vmovdqu  xmm0, qword ptr [rsp+1A8H]
       vmovdqu  qword ptr [rsp+E8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1B8H]
       vmovdqu  qword ptr [rsp+F8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1C8H]
       vmovdqu  qword ptr [rsp+108H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+1D8H]
       vmovdqu  qword ptr [rsp+118H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+168H]
       vmovdqu  qword ptr [rsp+A8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+178H]
       vmovdqu  qword ptr [rsp+B8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+188H]
       vmovdqu  qword ptr [rsp+C8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+198H]
       vmovdqu  qword ptr [rsp+D8H], xmm0
       mov      rcx, 0xD1FFAB1E
       call     CORINFO_HELP_NEWSFAST
       mov      rsi, rax
       mov      rcx, rsi
       xor      rdx, rdx
       call     AssertEqualityComparer`1:.ctor(ref):this
       vmovdqu  xmm0, qword ptr [rsp+E8H]
       vmovdqu  qword ptr [rsp+68H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+F8H]
       vmovdqu  qword ptr [rsp+78H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+108H]
       vmovdqu  qword ptr [rsp+88H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+118H]
       vmovdqu  qword ptr [rsp+98H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+A8H]
       vmovdqu  qword ptr [rsp+28H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+B8H]
       vmovdqu  qword ptr [rsp+38H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+C8H]
       vmovdqu  qword ptr [rsp+48H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+D8H]
       vmovdqu  qword ptr [rsp+58H], xmm0
G_M53037_IG03:
       lea      rcx, bword ptr [rsp+68H]
       lea      rdx, bword ptr [rsp+28H]
       mov      r8, rsi
       call     Assert:Equal(struct,struct,ref)
       nop      
G_M53037_IG04:
       add      rsp, 488
       pop      rsi
       pop      rdi
       ret      
; Total bytes of code 579, prolog size 35 for method Matrix4x4Tests:Matrix4x4NegateTest():this
; ============================================================
Unwind Info:
  >> Start offset   : 0x000000 (not in unwind data)
  >>   End offset   : 0xd1ffab1e (not in unwind data)
  Version           : 1
  Flags             : 0x00
  SizeOfProlog      : 0x09
  CountOfUnwindCodes: 4
  FrameRegister     : none (0)
  FrameOffset       : N/A (no FrameRegister) (Value=0)
  UnwindCodes       :
    CodeOffset: 0x09 UnwindOp: UWOP_ALLOC_LARGE (1)     OpInfo: 0 - Scaled small  
      Size: 61 * 8 = 488 = 0x001E8
    CodeOffset: 0x02 UnwindOp: UWOP_PUSH_NONVOL (0)     OpInfo: rsi (6)
    CodeOffset: 0x01 UnwindOp: UWOP_PUSH_NONVOL (0)     OpInfo: rdi (7)
```
and
```asm
; Assembly listing for method Matrix4x4Tests:Matrix4x4op_NegateTest():this
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;* V00 this         [V00    ] (  0,  0   )     ref  ->  zero-ref    this class-hnd
;  V01 loc0         [V01    ] (  3,  3   )  struct (64) [rsp+0x168]   do-not-enreg[XSB] must-init addr-exposed
;  V02 loc1         [V02    ] (  2,  2   )  struct (64) [rsp+0x128]   do-not-enreg[XSB] must-init addr-exposed
;  V03 OutArgs      [V03    ] (  1,  1   )  lclBlk (32) [rsp+0x00]  
;  V04 tmp1         [V04,T01] (  2,  4   )  struct (64) [rsp+0xE8]   do-not-enreg[SB]
;  V05 tmp2         [V05,T02] (  2,  4   )  struct (64) [rsp+0xA8]   do-not-enreg[SB]
;  V06 tmp3         [V06,T00] (  3,  6   )     ref  ->  rsi         class-hnd exact
;  V07 tmp4         [V07    ] (  4,  8   )  struct (64) [rsp+0x68]   do-not-enreg[XSB] addr-exposed
;  V08 tmp5         [V08,T03] (  2,  4   )    long  ->  rcx        
;  V09 tmp6         [V09    ] (  2,  4   )  struct (64) [rsp+0x28]   do-not-enreg[XSB] addr-exposed
;
; Lcl frame size = 424
G_M27564_IG01:
       push     rdi
       push     rsi
       sub      rsp, 424
       vzeroupper 
       mov      rsi, rcx
       lea      rdi, [rsp+128H]
       mov      ecx, 32
       xor      rax, rax
       rep stosd 
       mov      rcx, rsi
G_M27564_IG02:
       lea      rcx, bword ptr [rsp+168H]
       call     Matrix4x4Tests:GenerateMatrixNumberFrom1To16():struct
       lea      rcx, bword ptr [rsp+128H]
       vmovdqu  xmm0, qword ptr [rsp+168H]
       vmovdqu  qword ptr [rsp+68H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+178H]
       vmovdqu  qword ptr [rsp+78H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+188H]
       vmovdqu  qword ptr [rsp+88H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+198H]
       vmovdqu  qword ptr [rsp+98H], xmm0
       lea      rdx, bword ptr [rsp+68H]
       call     Matrix4x4:op_UnaryNegation(struct):struct
       vmovdqu  xmm0, qword ptr [rsp+168H]
       vmovdqu  qword ptr [rsp+E8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+178H]
       vmovdqu  qword ptr [rsp+F8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+188H]
       vmovdqu  qword ptr [rsp+108H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+198H]
       vmovdqu  qword ptr [rsp+118H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+128H]
       vmovdqu  qword ptr [rsp+A8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+138H]
       vmovdqu  qword ptr [rsp+B8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+148H]
       vmovdqu  qword ptr [rsp+C8H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+158H]
       vmovdqu  qword ptr [rsp+D8H], xmm0
       mov      rcx, 0xD1FFAB1E
       call     CORINFO_HELP_NEWSFAST
       mov      rsi, rax
       mov      rcx, rsi
       xor      rdx, rdx
       call     AssertEqualityComparer`1:.ctor(ref):this
       vmovdqu  xmm0, qword ptr [rsp+E8H]
       vmovdqu  qword ptr [rsp+68H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+F8H]
       vmovdqu  qword ptr [rsp+78H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+108H]
       vmovdqu  qword ptr [rsp+88H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+118H]
       vmovdqu  qword ptr [rsp+98H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+A8H]
       vmovdqu  qword ptr [rsp+28H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+B8H]
       vmovdqu  qword ptr [rsp+38H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+C8H]
       vmovdqu  qword ptr [rsp+48H], xmm0
       vmovdqu  xmm0, qword ptr [rsp+D8H]
       vmovdqu  qword ptr [rsp+58H], xmm0
       lea      rcx, bword ptr [rsp+68H]
       lea      rdx, bword ptr [rsp+28H]
       mov      r8, rsi
       call     Assert:Equal(struct,struct,ref)
       nop      
G_M27564_IG03:
       add      rsp, 424
       pop      rsi
       pop      rdi
       ret      
; Total bytes of code 499, prolog size 35 for method Matrix4x4Tests:Matrix4x4op_NegateTest():this
; ============================================================
Unwind Info:
  >> Start offset   : 0x000000 (not in unwind data)
  >>   End offset   : 0xd1ffab1e (not in unwind data)
  Version           : 1
  Flags             : 0x00
  SizeOfProlog      : 0x09
  CountOfUnwindCodes: 4
  FrameRegister     : none (0)
  FrameOffset       : N/A (no FrameRegister) (Value=0)
  UnwindCodes       :
    CodeOffset: 0x09 UnwindOp: UWOP_ALLOC_LARGE (1)     OpInfo: 0 - Scaled small  
      Size: 53 * 8 = 424 = 0x001A8
    CodeOffset: 0x02 UnwindOp: UWOP_PUSH_NONVOL (0)     OpInfo: rsi (6)
    CodeOffset: 0x01 UnwindOp: UWOP_PUSH_NONVOL (0)     OpInfo: rdi (7)
```

tannergooding replied:
Which is exactly what I expected.

@EgorBo, could you fix them up accordingly, since the JIT already does the right thing.

EgorBo replied:
@tannergooding um.. so how should I fix it and do not regress the software fallback (will it be inlined in `.Negate` if SSE is not available?). Is not the current implementation what you meant https://github.com/dotnet/corefx/blob/21ca6ac66e79ee99b3933660fe83ad1f2e07e545/src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs#L1855-L1881

tannergooding replied:
I'm not sure what you mean, in both cases `public static unsafe Matrix4x4 Negate(Matrix4x4 value) => -value;` will result in a call directly to `op_Negate`.

The JIT will then determine if `op_Negate` is appropriate to inline or not (currently it does not). It will result in code that is simpler to maintain and produces the same codegen as is being produced today (for Release builds)

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1949,219 +2050,104 @@ public static Matrix4x4 Multiply(Matrix4x4 value1, Matrix4x4 value2)
-
-            return m;
-        }
+        public static Matrix4x4 operator -(Matrix4x4 value) => Negate(value);
```
Nit: I think the normal case would be to have the implementation in the operator and have `Negate` call it (more people likely use the operator than call the "friendly name")

EgorBo commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1786,64 +1789,96 @@ public static Matrix4x4 Transpose(Matrix4x4 matrix)
-            result.M12 = matrix1.M12 + (matrix2.M12 - matrix1.M12) * amount;
-            result.M13 = matrix1.M13 + (matrix2.M13 - matrix1.M13) * amount;
-            result.M14 = matrix1.M14 + (matrix2.M14 - matrix1.M14) * amount;
+            if (Sse.IsSupported)
```
btw, initially I wanted to do `If (Sse.IsSupported) {} else if (Arm.Simd.IsSupported) {} else {}` 🙂

tannergooding replied:
Setting `COMPlus_FeatureSIMD=0` should cover the Sse-Avx2 HWIntrinsics. I don't think we have a switch to cover all HWIntrinsics, however.

CC. @CarolEidt

tannergooding replied:
The ARM intrinsics still need to be reviewed and implemented, before they can be used.

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -2149,12 +2300,23 @@ public static Matrix4x4 Multiply(Matrix4x4 value1, float value2)
+            if (Sse.IsSupported)
+            {
+                return
+                    Sse.MoveMask(Sse.CompareEqual(Sse.LoadVector128(&value1.M11), Sse.LoadVector128(&value2.M11))) != 0xF ||
```
`MoveMask(CompareNotEqual()) != 0` is probably more efficient

EgorBo replied:
just tested, you are right, it's 30% more efficient. Also, I tried to add some fast-out paths (compare first field or first row as simple m1.M11 == m2.M11 && ...) but it gave a minor improvement for some cases and major regression in anothers.

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1824,26 +1852,33 @@ public static Matrix4x4 Lerp(Matrix4x4 matrix1, Matrix4x4 matrix2, float amount)
         /// <returns>The negated matrix.</returns>
         public static Matrix4x4 Negate(Matrix4x4 value)
```
Not sure if you saw, since the thread is now "hidden" by GitHub, but I commented [here](https://github.com/dotnet/corefx/pull/31779#discussion_r210401298) (with disassembly using current head of CoreCLR) that the JIT does the right thing for inlining `public static Matrix4x4 Negate(Matrix4x4 value) => -value`

EgorBo replied:
yeah it's hard to follow when github collapses them 🙂

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -2305,10 +2305,10 @@ public static unsafe Matrix4x4 Multiply(Matrix4x4 value1, float value2)
-                    Sse.MoveMask(Sse.CompareEqual(Sse.LoadVector128(&value1.M21), Sse.LoadVector128(&value2.M21))) != 0xF ||
-                    Sse.MoveMask(Sse.CompareEqual(Sse.LoadVector128(&value1.M31), Sse.LoadVector128(&value2.M31))) != 0xF ||
-                    Sse.MoveMask(Sse.CompareEqual(Sse.LoadVector128(&value1.M41), Sse.LoadVector128(&value2.M41))) != 0xF;
+                    Sse.MoveMask(Sse.CompareNotEqual(Sse.LoadVector128(&value1.M11), Sse.LoadVector128(&value2.M11))) == 0 ||
```
Pretty sure this one should be `!= 0`. `CompareNotEqual` does `(a != b) ? 0xFFFFFFFF : 0`, so it will be non-zero if one of rows in `value1` != the same row in `value2`

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -2280,10 +2280,10 @@ public static unsafe Matrix4x4 Multiply(Matrix4x4 value1, float value2)
-                    Sse.MoveMask(Sse.CompareEqual(Sse.LoadVector128(&value1.M21), Sse.LoadVector128(&value2.M21))) == 0xF &&
-                    Sse.MoveMask(Sse.CompareEqual(Sse.LoadVector128(&value1.M31), Sse.LoadVector128(&value2.M31))) == 0xF &&
-                    Sse.MoveMask(Sse.CompareEqual(Sse.LoadVector128(&value1.M41), Sse.LoadVector128(&value2.M41))) == 0xF;
+                    Sse.MoveMask(Sse.CompareNotEqual(Sse.LoadVector128(&value1.M11), Sse.LoadVector128(&value2.M11))) != 0 &&
```
Pretty sure this one should be `== 0`. `CompareNotEqual` does `(a != b) ? 0xFFFFFFFF : 0`, so it will be zero if all of the rows in `value1` == the same row in `value2`

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -2346,9 +2181,9 @@ public bool Equals(Matrix4x4 other)
         public override bool Equals(object obj)
         {
-            if (obj is Matrix4x4)
+            if (obj is Matrix4x4 m)
```
Nit: You could do this in one line with `(obj is Matrix4x4 other) && (this == other)`

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1755,28 +1758,45 @@ public static Matrix4x4 Transform(Matrix4x4 value, Quaternion rotation)
+                var t2 = Sse.Shuffle(Sse.LoadVector128(&matrix.M11), Sse.LoadVector128(&matrix.M21), 0xEE);
+                var t3 = Sse.Shuffle(Sse.LoadVector128(&matrix.M31), Sse.LoadVector128(&matrix.M41), 0xEE);
+
+                Sse.Store(&result.M11, Sse.Shuffle(t0, t1, 0x88));
```
It would be nice to comment what `0x88` is (it looks like `ZXZX`)

The same for 0xDD (`WYWY`), 0x44 (`YXYX`), and 0xEE (`WZWZ`)

EgorBo replied:
turns out Unpack+Move works better and no magic numbers are involved 🙂

eerhardt commented:
@EgorBo - the CI is failing because `System.Numerics.Vectors` is built for:

* netcoreapp
* netfx
* netstandard

The Hardware Intrinsics methods you are taking advantage of (i.e. `Vector128<>`) only work/exist on `netcoreapp`. So your code builds fine on that build.

However, for netfx and netstandard, `System.Runtime.Intrinsics` doesn't exist.  So you will need to exclude your new code from the build when building for these configurations.

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1779,15 +1806,31 @@ public static Matrix4x4 Transpose(Matrix4x4 matrix)
         }
 
+#if netcoreapp
+        private static Vector128<float> Lerp(Vector128<float> a, Vector128<float> b, Vector128<float> t) => 
```
It would be nice if this was in a new `internal static class VectorMath` type, to facilitate sharing with `Vector2/3/4` (and the other types, where possible)

EgorBo replied:
I wrote a small prototype of a such VectorMath but I am not sure about the fallbacks, could you please take a look? https://gist.github.com/EgorBo/6e0c1d78ee07cfc3fcb52cd3df7d9120

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -3,12 +3,18 @@
 
 using System.Globalization;
+using System.Runtime.InteropServices;
+#if netcoreapp
```
This is currently not "working" as expected. I don't see `netcoreapp` in the `DefineConstants` list

tannergooding commented on src/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj:
```diff
@@ -4,6 +4,7 @@
     <RootNamespace>System.Numerics</RootNamespace>
     <DocumentationFile>$(OutputPath)$(MSBuildProjectName).xml</DocumentationFile>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <DefineConstants Condition="'$(TargetGroup)'=='netcoreapp'">$(DefineConstants);netcoreapp</DefineConstants>
```
@eerhardt, could you confirm if this is the desired way to do this?

eerhardt replied:
Typically we *try* to avoid `#if` and `DefineConstants` in our code, and instead opt for whole files that are included for each target framework.  However, in this case, I don't see a good way to do that unfortunately.  (If you can find a good way, feel free to use that approach.)

If we do need to use `DefineConstants`, it is preferred to use constants that are tied to the concept that you need, and not which target happens to have that capability today.  For example, if you need to use the Registry, you'd have a constant for `HAS_REGISTRY` or `FEATURE_REGISTRY`, instead of having `#if WINDOWS`.

So in this case `HAS_INTRINSICS` seems more preferred, than `NETCOREAPP`.  See https://github.com/dotnet/corefx/commit/c3c2faac6f691a3f1520e616794da0af630b7d51#diff-a50daad0581f9c1c7d0d39fc82cd9457L304 for an example where we are using Span.

EgorBo replied:
@eerhardt  Got it! Are intrinsics available for CoreRT?

tannergooding replied:
Not today.

tannergooding commented:
In case people are interested, here are the current PMI disassemblies:
* Software Implementation: [System.Numerics.Vectors.dasm.txt](https://github.com/dotnet/corefx/files/2294452/System.Numerics.Vectors.dasm.txt)
* Hardware Intrinsics Impl: [System.Numerics.Vectors.dasm.txt](https://github.com/dotnet/corefx/files/2294453/System.Numerics.Vectors.dasm.txt)

CarolEidt commented:
> Setting COMPlus_FeatureSIMD=0 should cover the Sse-Avx2 HWIntrinsics. I don't think we have a switch to cover all HWIntrinsics, however.

Would it make sense to have a `COMPlus_EnableHWIntrinsics` option that is exposed in Release? It would be set to 1 by default, and could be set to 0 to disable them. Presumably, it would imply `COMPlus_FeatureSIMD` and perhaps the latter should be deprecated?

tannergooding commented:
> Would it make sense to have a COMPlus_EnableHWIntrinsics option that is exposed in Release? It would be set to 1 by default, and could be set to 0 to disable them. Presumably, it would imply COMPlus_FeatureSIMD and perhaps the latter should be deprecated?

I think this makes sense, but we probably don't want to deprecate `FeatureSIMD` until all of the types are successfully ported (Matrix4x4 is a case where nothing was Intrinsic, `Vector2/3/4` are the ones that `FeatureSIMD` impacts, IIRC).

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/VectorMath.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            if (Sse.IsSupported)
+            {
+                // do not use !Equal if vectors contain NaN
```
You probably should use `CompareEqual` and `0xF` here (for `==`) since it properly handles `NaN` (reports `false` for `NaN`)

You can use `CompareNotEqual` for `!=`, where it properly handles `NaN` (reports `true` for `NaN`)

tannergooding replied:
@EgorBo, could you provide an example of where `Sse.CompareNotEqual(vector1, vector2)) != 0` fails for `!=`?

EgorBo replied:
![image](https://user-images.githubusercontent.com/523221/44238988-bb6b5880-a1bf-11e8-9491-209786a2e7e7.png)

tannergooding replied:
How is that incorrect? `v1 != v2` so `notEqual is true`

EgorBo replied:
@tannergooding sorry, wrong example. I meant {NaN,0,0,0} != {NaN,0,0,0}, the code above will return `true` but there is a test: https://github.com/dotnet/corefx/blob/master/src/System.Numerics.Vectors/tests/Matrix4x4Tests.cs#L2438-L2439

EgorBo replied:
@tannergooding now I get it. for some reason I was convinced that when NaN is involved anywhere - both == and != should always be false.

tannergooding replied:
Ah. Yeah, for `NaN`:
* == is always false
* != is always true

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1755,8 +1760,29 @@ public static Matrix4x4 Transform(Matrix4x4 value, Quaternion rotation)
+#if HAS_INTRINSICS
+            if (Sse.IsSupported)
+            {
+                var row1 = Sse.LoadVector128(&matrix.M11);
```
@CarolEidt, @fiigii:

The codegen for this is pretty sub-optimal:
```asm
       mov      rax, rdx
       vmovups  xmm0, xmmword ptr [rax]
       lea      rax, bword ptr [rdx+16]
       mov      r8, rax
       vmovups  xmm1, xmmword ptr [r8]
       lea      r8, bword ptr [rdx+32]
       mov      r9, r8
       vmovups  xmm2, xmmword ptr [r9]
       lea      r9, bword ptr [rdx+48]
       mov      r10, r9
       vmovups  xmm3, xmmword ptr [r10]
```

tannergooding replied:
Logged https://github.com/dotnet/coreclr/issues/19521 to track this one

stephentoub replied:
Nit: `var` => `Vector128<Matrix4x4>`, assuming that's what it is.  Same goes for other `var` usage... we only use `var` when the type is fully specified on the RHS.

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1755,8 +1760,29 @@ public static Matrix4x4 Transform(Matrix4x4 value, Quaternion rotation)
+                var h12 = Sse.UnpackHigh(row1, row2);
+                var h34 = Sse.UnpackHigh(row3, row4);
+
+                Sse.Store(&matrix.M11, Sse.MoveLowToHigh(l12, l34));
```
nevermind, you can't fold these stores 😄

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -1755,8 +1760,29 @@ public static Matrix4x4 Transform(Matrix4x4 value, Quaternion rotation)
+                Sse.Store(&matrix.M31, Sse.MoveLowToHigh(h12, h34));
+                Sse.Store(&matrix.M41, Sse.MoveHighToLow(h34, h12));
+
+                return matrix;
```
Logged https://github.com/dotnet/coreclr/issues/19522

tannergooding commented on src/System.Numerics.Vectors/src/System/Numerics/VectorMath.cs:
```diff
@@ -0,0 +1,38 @@
+    {
+        public static Vector128<float> Lerp(Vector128<float> a, Vector128<float> b, Vector128<float> t)
+        {
+            if (Sse.IsSupported)
```
It's probably sufficient to just `Debug.Assert(Sse.IsSupported)`, since this is internal and we have already checked the code higher up.

fiigii commented:
IMO, the current implementation has a very inefficient code pattern that **loads** fields of `Matrix4x4` at the beginning of a function and **stores** the results back at the end. Each operation could generate many memory access operations, so some scenarios that heavily use `Matrix4x4` will possibly become memory bounded.

So, the more efficient implementation should be defining `Matrix4x4` with 4 `Vector128<float>` fields, and other operations directly read the `Vector128<float>` fields then return a `new Matrix4x4(...)`. Because of RyuJIT's struct promotion optimization and inlining, all the loads of local struct's fields and "new" of results will be eliminated (replaced by direct SIMD register operations). According to my experience, that will make some scenarios much faster and more CPU bounded.

cc @tannergooding @CarolEidt

EgorBo commented:
@fiigii how would a field accessor look like (e.g. `M11`) - both setter and getter? Will it be interop friendly?

tannergooding commented:
@fiigii, I don't believe we can take such a change because the various `System.Numerics.Vector` types were shipped with their fields publicly exposed.

We cannot:
* Remove the existing fields, as it is a major binary breaking change
* Add new fields with explicit layout, as it would
  * Break any existing code using these as interop types (generics can't be marshaled today)
  * Likely regress perf in other areas, due to JIT limitations with explicit layout structs

This also prevents other changes that likely would have been generally good for these types, such as marking the structs as `readonly`, but it is the state of the world and we will need to work with it 😄

tannergooding commented:
I would hope that, instead, we could make the JIT smarter for these types of things. Having `__vectorcall` support, for example, would allow the values to be passed around in register and would elide many of the memory concerns.

tannergooding commented:
@EgorBo, are there any other changes you were looking to make to this PR?

EgorBo commented:
@tannergooding let me switch back to Shuffle-impl for Multiply 🙂

eerhardt commented on src/System.Numerics.Vectors/src/System/Numerics/VectorMath.cs:
```diff
@@ -0,0 +1,32 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+#if HAS_INTRINSICS
```
Instead of completely `#if` on this whole file, we should only include it when it is used.

So, we can remove this `#if`, and in the .csproj add it to `<Compile>` when `$(TargetsNetCoreApp)`.

EgorBo replied:
@eerhardt yeah I was thinking about it, but I guess I have to declare a new property $(IntrinsicsSupport) something like this in the csproj since we rely on HAS_INTRINSICS symbol

eerhardt commented on src/System.Numerics.Vectors/src/System/Numerics/VectorMath.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    internal static class VectorMath
+    {
+        public static Vector128<float> Lerp(Vector128<float> a, Vector128<float> b, Vector128<float> t)
```
I've seen guidance on other PRs that putting `[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]` on functions that take/return Vector128 is a good idea.

See [https://github.com/briancylui/machinelearning/issues/3#issuecomment-410118363](https://github.com/briancylui/machinelearning/issues/3#issuecomment-410118363) where we were seeing perf degradation when we refactored some "helper" vector methods.

tannergooding replied:
It can help, but looking at the existing codegen it looks good/correct as is.

tannergooding replied:
(that is, the method is small enough to be automatically inlined)

EgorBo replied:
Yeah I've just checked both outputs - both Lerp and Equals\NotEquals are inlined.

eerhardt replied:
OK, thanks for checking. It's good to know that this refactoring didn't hurt perf.

EgorBo commented:
I've updated benchmark results in the description with the latest changes, latest runtime and added macOS.
macOS numbers a little bit worse because most of the used intrinsics have better throughput on Coffee Lake (e.g. https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_add_ps&expand=127)

CarolEidt commented:
@EgorBo - thanks! This is great to see.
@tannergooding @eerhardt - thanks for getting this reviewed and merged. I've commented on a couple of the JIT issues that were raised. Let me know if there are others.
